### PR TITLE
CNI: add host-side interface info to cni.Result

### LINKED ISSUE #24899: cilium CNI cannot be used in combination with `bandwidth` CNI
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

In my testing environment, I tested the scenario of using `cilium` and `bandwidth` CNI together(we cannot use cilium `bandwidthManager` because our os kernel version is too low to meet the requirements.). Then we found that an error occured when `bandwith` CNI was called: `no veth peer of container interface found in host ns`.

After reivewing the `bandwith` CNI source code, the reason for this error is that the result returned by cilium CNI does not meet the following containtion.
https://github.com/containernetworking/plugins/blob/4a6147a1552064af80b4f7567b30c5174153c62a/plugins/meta/bandwidth/main.go#L153-L160
```go
for _, iface := range interfaces {
    if iface.Sandbox == "" && iface.Name == link.Attrs().Name {
	return iface, nil
    }
}
```

The official documentation of CNI requires that the `interfaces` field should include information about both container-side interface and host-side interface. However, currently cilium only provides the container-side interface information.

https://www.cni.dev/docs/spec/#section-5-result-types
- `interfaces`: An array of all interfaces created by the attachment, including any host-level interfaces:
    - `name`: The name of the interface.
    - `mac`: The hardware address of the interface (if applicable).
    - `sandbox`: The isolation domain reference (e.g. path to network namespace) for the interface, or empty if on the host. For interfaces created inside the container, this should be the value passed via `CNI_NETNS`.

some test code:
```diff
+       log.Println("interfaces length:", len(interfaces))
        for _, iface := range interfaces {
+               log.Println("iface name:", iface.Name)
+               log.Println("sandbox: ", iface.Sandbox)
+               log.Println("link.Attrs().Name: ", link.Attrs().Name)
                if iface.Sandbox == "" && iface.Name == link.Attrs().Name {
                        return iface, nil
                }
        }
```

output:
```
2023/04/13 07:20:51 interfaces length: 1
2023/04/13 07:20:51 iface name: eth0
2023/04/13 07:20:51 sandbox:  /var/run/netns/cni-882cb34f-9dd9-6014-556b-724e5d2e383d
2023/04/13 07:20:51 link.Attrs().Name:  lxc1986c9428372
```

### Cilium Version

1.13.0

### Kernel Version

5.15.0-69-generic

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:40:17Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:33:12Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by nayihz**
Fixes: https://github.com/cilium/cilium/issues/24899

```release-note
Add host-side interface info to cni.Result, which allows bandwidth CNI to work with Cilium
```

### CODE REVIEW

**Comment by tommyp1ckles** [comment]
@czybjtu thanks for the PR 😄 

One question, later on in the function IPs get their interface index added: https://github.com/cilium/cilium/blob/5b042476dd6f18c70bc7c0d7401cc596abf04b24/plugins/cilium-cni/main.go#L608

Would this still work if we add the host level interface first?

**Comment by czybjtu** [comment]
> Would this still work if we add the host level interface first?

Thanks for catching this. Have fixed it. PTAL.

File: plugins/cilium-cni/main.go [hunk 1]
```diff
@@ -552,6 +557,9 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 			err = fmt.Errorf("unable to prepare IP addressing for '%s': %s", ep.Addressing.IPV6, err)
 			return
 		}
+		// Add to the result the Interface as index of Interfaces
+		ipConfig.Interface = cniTypesV1.Int(len(res.Interfaces))
```
Comment (tommyp1ckles) [inline plugins/cilium-cni/main.go:543]: Shouldn't this be:
```suggestion
		ipConfig.Interface = cniTypesV1.Int(len(res.Interfaces)-1)
```

Comment (nayihz) [reply plugins/cilium-cni/main.go:543]: You cannot do it this way because the index of an array starts from 0.

Comment (tommyp1ckles) [reply plugins/cilium-cni/main.go:543]: Sorry I'm still confused,

On line: https://github.com/cilium/cilium/pull/24901/files#diff-87793c7d9c6d6a10f34d6a3efefed53dac8abd700cb05cfb2a77844ec990b744R523

So at this point res.Interfaces is either of length 0 or 1, if ip4 and ip6 are enabled then we're going to be referencing 2, which is out of bounds in the Interfaces array?

Does that make sense? This function is a bit confusing so I might be missing something here.

Comment (nayihz) [reply plugins/cilium-cni/main.go:543]: On line L523, we add the host-side veth info to `res.Interface` which has no ip.
On line L561 and L577, `ipConfig` was refer to container-side veth info. `ipConfig` is the **last element** that will be added to `res.Interface` slice. So I think
```go
// ipConfig.Interface represents its index in this slice.
ipConfig.Interface = cniTypesV1.Int(len(res.Interfaces))
```
is what we want.

**Comment by joestringer** [comment]
I've kicked off the tests to ensure there are no regressions introduced by this. Other than that I think we are just looking for @tommyp1ckles' feedback on the remaining open thread.

File: plugins/cilium-cni/main.go (hunk 1 above)
Comment (tommyp1ckles) [reply plugins/cilium-cni/main.go:543]: Does that still work if datapathmodeveth isn't enabled?

(Line 504)

Comment (nayihz) [reply plugins/cilium-cni/main.go:543]: Yes. We add host side veth info to `res.Interfaces` only when datapathmodeveth enabled. Everything is the same as before if datapathmodeveth was not enabled.

Comment (tommyp1ckles) [reply plugins/cilium-cni/main.go:543]: Ok I see now, thanks for the clarification.

**Comment by tommyp1ckles** [comment]
/test-1.25-4.19

**Comment by tommyp1ckles** [comment]
/test-1.26-net-next

**Comment by tommyp1ckles** [comment]
/test-1.16-4.19

**Comment by tommyp1ckles** [comment]
/test-1.24-5.4

**Comment by tommyp1ckles** [comment]
Changes LGTM 🙏 , just gonna make sure tests are passing

**Comment by tommyp1ckles** [comment]
@czybjtu sorry, just getting back to this one, do you mind rebasing and we can make sure the tests pass.

**Review by tommyp1ckles — APPROVED** [review]
Looks good, just going to make sure tests pass

**Comment by czybjtu** [comment]
```
❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 --retry 3 --retry-all-errors --retry-delay 3 [http://google.com:80](http://google.com/)" failed: command failed: curl: (28) Resolving timed out after 2000 milliseconds
```
This doesn't seem to be related to this PR.

**Comment by ti-mo** [comment]
I've approved the tests to run. Could you please add a commit description with some of the context you've outlined in the original issue? right now it's hard to tell what the intention is behind the change by just looking at the commit.

**Comment by czybjtu** [comment]
Updated commit message. PTAL.

**Comment by tommyp1ckles** [comment]
not sure why tests failed, possible GH Actions issue 😕 

/test

**Comment by ldelossa** [comment]
@czybjtu GH actions is looking really off here. Can you rebase this PR and force push?

**Comment by czybjtu** [comment]
I have some problems when rebase upstream because the branch name has changed(master -> main). I will open a new same pr for this.
Sorry for inconvenience.

**Comment by czybjtu** [comment]
https://github.com/cilium/cilium/pull/26518

New pr has been created. PTAL

**Pull request closed without merging**